bundles/org.openhab.binding.doorbird/src/main/java/org/openhab/binding/doorbird/internal/DoorbirdUdpListener.java

```diff
@@ -0,0 +1,151 @@
+    private @Nullable DatagramSocket socket;
+
+    private byte @Nullable [] lastData;
+    private int lastDataLength;
+    private long lastDataTime;
+
+    public DoorbirdUdpListener(DoorbirdHandler thingHandler) {
+        this.thingHandler = thingHandler;
+    }
+
+    @Override
+    public void run() {
+        receivePackets();
+    }
+
+    public void shutdown() {
+        if (socket != null) {
+            socket.close();
+            logger.debug("Listener closing listener socket");
+            socket = null;
+        }
+    }
+
+    private void receivePackets() {
+        try {
+            socket = new DatagramSocket(UDP_PORT);
+            socket.setSoTimeout(SOCKET_TIMEOUT);
+            logger.debug("Listener got UDP socket on port {} with timeout {}", UDP_PORT, SOCKET_TIMEOUT);
+        } catch (SocketException e) {
+            logger.debug("Listener got SocketException: {}", e.getMessage(), e);
+            socket = null;
+            return;
+        }
+
+        DatagramPacket packet = new DatagramPacket(new byte[BUFFER_SIZE], BUFFER_SIZE);
+        while (socket != null) {
+            try {
+                socket.receive(packet);
+                processPacket(packet);
+            } catch (SocketTimeoutException e) {
+                // Nothing to do on socket timeout
+            } catch (IOException e) {
+                logger.debug("Listener got IOException waiting for datagram: {}", e.getMessage());
+                socket = null;
+            }
+        }
+        logger.debug("Listener exiting");
+    }
+
+    private void processPacket(DatagramPacket packet) {
+        logger.trace("Got datagram of length {} from {}", packet.getLength(), packet.getAddress().getHostAddress());
+
+        // Check for duplicate packet
+        if (isDuplicate(packet)) {
+            logger.trace("Dropping duplicate packet");
+            return;
+        }
+
+        try {
+            String userPassword = thingHandler.getUserPassword();
+            if (userPassword != null) {
+                event.decrypt(packet, userPassword);
+            }
+        } catch (RuntimeException e) {
+            // The libsodium library might generate a runtime exception if the packet is malformed
+            logger.info("DoorbirdEvent got unhandled exception: {}", e.getMessage(), e);
+            return;
+        }
+
+        String doorbirdId = thingHandler.getDoorbirdId();
+        if (event.isDoorbellEvent() && doorbirdId != null) {
+            if ("motion".equalsIgnoreCase(event.getEventId())) {
+                thingHandler.updateMotionChannel(event.getTimestamp());
+            } else if (doorbirdId.equalsIgnoreCase(event.getEventId())) {
+                thingHandler.updateDoorbellChannel(event.getTimestamp());
+            } else {
+                logger.debug("Unknown doorbell event type: {}", event.getEventId());
+            }
+        }
+    }
+
+    private boolean isDuplicate(DatagramPacket packet) {
```

Comment:
Why would we receive duplicate packets? Does doorbird send them?

Reply:
Yes, it sends at least 2 packets per user defined in the Doorbird app. So, for example, if you have 3 users defined in the app, you'll get at least 6 packets.